components/datetime/src/skeleton.rs

```diff
@@ -260,10 +287,339 @@ impl From<fields::SymbolError> for SkeletonError {
+/// search for a best skeleton.
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum BestSkeleton<'a> {
+    AllFieldsMatch(AvailableFormatPattern<'a>),
+    MissingOrExtraFields(AvailableFormatPattern<'a>),
+    NoMatch,
+}
+
+/// A partial implementation of the [UTS 35 skeleton matching algorithm](https://unicode.org/reports/tr35/tr35-dates.html#Matching_Skeletons).
+///
+/// The following is implemented:
+///
+///  * Compute a score based on the best possible match for the given fields.
+///  * Select the skeleton with highest score.
+///
+/// The following is not implemented:
+///
+///  * 2.6.2.1 Matching Skeletons
+///    - TODO(#584) - Modify the resulting pattern to have fields of the same length. For example requesting
+///      a skeleton "yMMMMd" can have a best match of ["yMMMd", "d MMM y"]. This pattern should
+///      then be modified to use the requested length to produce a pattern "d MMMM y".
+///      However, fields should not be changed from numeric to text.
+///  * 2.6.2.2 Missing Skeleton Fields
+///    - TODO(#585) - The mechanism to combine a date pattern and a time pattern.
+///    - TODO(#586) - Using the CLDR appendItems field. Note: There is not agreement yet on how
+///      much of this step to implement. See the issue for more information.
+pub fn get_best_available_format_pattern<'a>(
+    available_format_patterns: impl Iterator<Item = AvailableFormatPattern<'a>> + 'a,
+    fields: &[Field],
+) -> BestSkeleton<'a> {
+    let mut closest_format_pattern = None;
+    let mut closest_distance: u32 = u32::MAX;
+    let mut closest_missing_fields = 0;
+
+    for available_format_pattern in available_format_patterns {
+        let skeleton = &available_format_pattern.skeleton;
+        debug_assert!(
+            skeleton.fields_len() <= MAX_SKELETON_FIELDS as usize,
+            "The distance mechanism assumes skeletons are less than MAX_SKELETON_FIELDS in length."
+        );
+        let mut missing_fields = 0;
+        let mut distance: u32 = 0;
+        // The distance should fit into a u32.
+
+        let mut requested_fields = fields.iter().peekable();
+        let mut skeleton_fields = skeleton.fields_iter().peekable();
+        loop {
+            let next = (requested_fields.peek(), skeleton_fields.peek());
+
+            // Try to find matching symbols.
+            match next {
+                (Some(requested_field), Some(skeleton_field)) => {
+                    debug_assert!(
+                        // As of the time of this writing, stand-alone months are not in the CLDR
+                        // skeleton data. The components::Bag could produce stand-alone month fields,
+                        // but since the CLDR does not have them, only Month::Format symbols are
+                        // used for matching.
+                        skeleton_field.symbol != FieldSymbol::Month(fields::Month::StandAlone)
+                    );
```

Comment:
Thought: Data could come from untrusted sources and have fields that don't exist in CLDR.  But, this is OK as a `debug_assert!`.  Just not a real `assert!`.

Reply:
Ah, that makes sense.